xwiki-rendering-macros/xwiki-rendering-macro-content/src/main/java/org/xwiki/rendering/internal/macro/content/ContentMacro.java

```diff
@@ -78,52 +84,45 @@ public class ContentMacro extends AbstractMacro<ContentMacroParameters>
     }
 
     @Override
     public boolean supportsInlineMode()
     {
-        return false;
+        return true;
     }
 
     @Override
-    public List<Block> execute(ContentMacroParameters parameters, String content, MacroTransformationContext context)
-        throws MacroExecutionException
+    public List<Block> execute(ContentMacroParameters parameters, String macroContent,
+        MacroTransformationContext context) throws MacroExecutionException
     {
-        try {
-            List<Block> blocks = getSyntaxParser(parameters.getSyntax()).parse(new StringReader(content)).getChildren();
-            MetaDataBlock metaDataBlock = new MetaDataBlock(blocks, MetaData.SYNTAX, parameters.getSyntax());
-
-            metaDataBlock.getMetaData().addMetaData(this.getNonGeneratedContentMetaData());
-            return Collections.singletonList(metaDataBlock);
-        } catch (ParseException e) {
-            throw new MacroExecutionException(
-                String.format("Failed to parse macro content in syntax [%s]", parameters.getSyntax()), e);
+        Syntax syntax = parameters.getSyntax();
+        String content = macroContent;
+        MetaData metadata = new MetaData();
+        if (syntax != null) {
+            // Remember the custom syntax as it can be needed for sub macros
+            metadata.addMetaData(MetaData.SYNTAX, syntax);
         }
-    }
-
-    /**
-     * Get the parser for the passed Syntax.
-     *
-     * @param syntax the Syntax for which to find the Parser
-     * @return the matching Parser that can be used to parse content in the passed Syntax
-     * @throws MacroExecutionException if there's no Parser in the system for the passed Syntax
-     */
-    protected Parser getSyntaxParser(Syntax syntax) throws MacroExecutionException
-    {
-        ComponentManager componentManager = this.componentManagerProvider.get();
-        if (componentManager.hasComponent(Parser.class, syntax.toIdString())) {
-            try {
-                return componentManager.getInstance(Parser.class, syntax.toIdString());
-            } catch (ComponentLookupException e) {
-                throw new MacroExecutionException(
-                    String.format("Failed to lookup Parser for syntax [%s]", syntax.toIdString()), e);
+        if (parameters.getSource() != null) {
+            MacroContentWikiSource wikiSource = this.contentFactory.getContent(parameters.getSource(), context);
+            if (parameters.getSyntax() == null) {
+                // Use the source syntax if no explicit one is given as parameter
+                syntax = wikiSource.getSyntax();
             }
+            content = wikiSource.getContent();
         } else {
-            throw new MacroExecutionException(String.format("Cannot find Parser for syntax [%s]", syntax.toIdString()));
+            // Make the content editable inline only if it's not coming from a source
+            metadata.addMetaData(this.getNonGeneratedContentMetaData());
```

Comment:
> At least I would be surprised if XWiki handled this correctly when the syntax is different.

I feel that would be something to fix in the XHTML parser, since it has everything needed to know the syntax.

Reply:
Okay, I've just had a look at the code and it seems like it most likely already handles that correctly, nice.

Reply:
I think what I remembered was that there is no way to determine programmatically which syntax a macro's content uses but apparently this issue doesn't exist for non-generated content.